src/status_emitter/json.rs

```diff
@@ -0,0 +1,157 @@
+}
+
+fn emit_test_end(name: &String, status: &String, error_output: &String) {
+    let escaped_name = serde_json::to_string(name).unwrap();
+    let escaped_error_output = if error_output.is_empty() {
+        String::new()
+    } else {
+        format!(r#", "stdout": {}"#, serde_json::to_string(error_output).unwrap())
+    };
+
+    // Adapted from test::formatters::json::write_event().
+    println!(r#"{{ "type": "test", "event": "{status}", "name": {escaped_name}{escaped_error_output} }}"#);
+}
+
+fn emit_test_start(name: &String) {
+    let escaped_name = serde_json::to_string(name).unwrap();
+
+    // Adapted from test::formatters::json::write_test_start().
+    println!(r#"{{ "type": "test", "event": "started", "name": {escaped_name} }}"#);
+}
+
+// MAINTENANCE REGION END
+
+/// A JSON output emitter.
+#[derive(Clone)]
+pub struct JSON;
+
+impl StatusEmitter for JSON {
+    /// Create a report about the entire test run at the end.
+    fn finalize(
+        &self,
+        failed: usize,
+        succeeded: usize,
+        ignored: usize,
+        filtered: usize,
+        aborted: bool,
+    ) -> Box<dyn Summary> {
+        let status = if aborted || failed > 0 {
+            String::from("failed")
+        } else {
+            String::from("ok")
+        };
+
+        emit_suite_end(failed, filtered, ignored, succeeded, &status);
+
+        Box::new(())
+    }
+
+    /// Invoked the moment we know a test will later be run.
+    /// Useful for progress bars and such.
+    fn register_test(&self, path: PathBuf) -> Box<dyn TestStatus + 'static> {
+        let name = path.to_str().unwrap().to_string();
+
+        emit_test_start(&name);
+
+        Box::new(JSONStatus {
+            name,
+            path,
+            revision: String::new(),
+            style: RevisionStyle::Show,
+        })
+    }
+}
+
+/// Information about a specific test run.
+pub struct JSONStatus {
+    name: String,
+    path: PathBuf,
+    revision: String,
+    style: RevisionStyle,
+}
+
+impl TestStatus for JSONStatus {
+    /// A test has finished, handle the result immediately.
+    fn done(&self, result: &TestResult, aborted: bool) {
+        let status = if aborted {
+            String::from("timeout")
+        } else {
+            match result {
+                Ok(TestOk::Ignored) => String::from("ignored"),
+                Ok(TestOk::Ok) => String::from("ok"),
+                Err(_) => String::from("failed"),
+            }
+        };
+        let error_output = if let Err(errored) = result {
+            let command = &errored.command;
+            let stderr = errored.stderr.to_str_lossy();
+            let stdout = errored.stdout.to_str_lossy();
+            format!(r#"---- command\n\n{command}\n\n\---- stdout\n\n\{stdout}\n\n---- stderr\n\n\{stderr}"#)
+        } else {
+            String::new()
+        };
+
+        emit_test_end(&self.name, &status, &error_output);
```

Comment:
the revision and the path are not part of the json, so you are getting duplicate messages. I'm ok with putting all of this into the name like `format!("{name}({revision}): {path}")` or sth. Usually when the revision is an empty string we don't print it at all (like not even empty parentheses). 

My preferred solution would be extra fields in the json, but I understand that this is suboptimal due to none of the json tooling consuming revision and path information yet.

Reply:
Note that I treat the path of the test as the name of the test. In addition, when creating a `TestStatus`, the revision is an empty string.

Is there a better way of obtaining the name and revision of a test?

Reply:
You are already tracking it correctly in `for_revision and `for_path. So I think you just need to print it as part of the name